Searchfield for help

nicodh opened the pull request:
This PR adds a search field to the help page, using [pagefind](https://pagefind.app) which generates an index which is then used to search and disply results in the static help page.

Needs latest main of https://github.com/deltachat/deltachat-pages/

![image](https://github.com/user-attachments/assets/66542aef-6a21-449f-8bd1-500978f1d80c)

![image](https://github.com/user-attachments/assets/c0c967ac-1279-4a87-b847-9b61851c0a2e)

Simon-Laux reviewed (commented):
Nice. For the translation keys I think it makes more sense to use the existing substitution pattern, you can convert it to the custom one in `bin/help/help-translations.js`.
- reasons: reusability and consistency

Simon-Laux commented on _locales/_untranslated_en.json:
```diff
@@ -44,5 +44,8 @@
   "multidevice_connection_takes_too_long": {
     "message": "This takes longer than expected 🐌 make sure both devices are on the same Wi-Fi or network and check troubleshooting"
+  },
+  "one_result":  {
```
the key says nothing about search or search term

nicodh replied:
changed the wording according to the existing keys

Simon-Laux commented on bin/help/create-local-help.sh:
```diff
@@ -0,0 +1,30 @@
+    rm ./static/help/$LANGUAGE/pagefind/*.css
+    rm ./static/help/$LANGUAGE/pagefind/pagefind-*.js
+  else
+    echo "Datei $FILE nicht gefunden"
```
english please
```suggestion
    echo "File $FILE not found"
```

ralphtheninja commented on static/help/cs/pagefind/pagefind.js:
```diff
@@ -0,0 +1,9 @@
+const pagefind_version="1.3.0";let wasm_bindgen;(function(){const __exports={};let script_src;if(typeof document!=='undefined'&&document.currentScript!==null){script_src=new URL("UNHANDLED",location.href).toString()}let wasm=undefined;let cachedUint8Memory0=null;function getUint8Memory0(){if(cachedUint8Memory0===null||cachedUint8Memory0.byteLength===0){cachedUint8Memory0=new Uint8Array(wasm.memor ...
```
These `static/help/${language}/pagefind/pagefind.js` are identical?

nicodh replied:
Yes, but the problem is, that there is no way (or I didn't find one) to tell the generated Javascript via config to use another path to access the generated fragments. So we would have to either change the minified Javascript or add someting to the pagefind project to enable our use case without file duplication.

Of course in Unix systems a symlink would be fine, but that causes problems on Windows/Mac I assume

ralphtheninja replied:
I'm fine with it. I was just wondering if it was expected or not. We solve the windows problem with redundancy.

Simon-Laux commented on static/help/en/help.html:
```diff
@@ -1,14 +1,14 @@
 <!DOCTYPE html>
-<html><head><meta charset="UTF-8" /><meta name="viewport" content="initial-scale=1.0" /><link rel="stylesheet" href="../help.css" /></head><body><ul id="top">
+<html><head><meta charset="UTF-8" /><meta name="viewport" content="initial-scale=1.0" /><link rel="stylesheet" href="../help.css" /><script src="./pagefind/pagefind-ui.js"></script><script src="../start-pagefind.js"></script></head><body><ul id="top">
```
"pagefind-ui.js" is wrong, it is just "pagefind.js"

Simon-Laux reviewed (changes requested):
does not work out of the box and I agree with the idea that `pagefind.js` and `wasm.unknown.pagefind` should not be duplicated - maybe they could be imported via import statement from `start-pagefind.js`?

Simon-Laux commented:
maybe `https://pagefind.app/docs/ui/#bundle-path` could help, Pagefind (the lib not ui) also has a `basePath` option

nicodh commented:
> does not work out of the box and I agree with the idea that `pagefind.js` and `wasm.unknown.pagefind` should not be duplicated - maybe they could be imported via import statement from `start-pagefind.js`?

Sorry -  the problem was, that I didn't commit the changes needed after the update of deltachat-pages

nicodh commented:
Concerning the redundant files: pagefind expects the wasm files to be at the same path as pagefind-ui.js & pagefind.js.
All together build the "bundle-path". The paths to the data are obviously relative.

What I did now is: copy the needed files to static/help/ and remove them from /static/help/<locale>/pagefind/ in build process.
Then copy them to the current language dir on runtime when needed.

Simon-Laux commented:
> Then copy them to the current language dir on runtime when needed. (when language is switched)

bad idea, app files at runtime should be treated immutable. what you can do is making a custom uri schema for the help and match on the name to return the same file in all sub directories, similar to how `webxdc.js` is served.

nicodh commented:
I found a solution: added a lang attribute to each help page and moved the pagefind indexing one level up, so all data goes into one pagefind directory while pagefind assigns the index per language and shows only appropriate matches

Simon-Laux commented:
<img width="486" alt="Bildschirmfoto 2025-02-27 um 14 47 39" src="https://github.com/user-attachments/assets/f0c7293e-ea76-4771-8809-7e0d954634b0" />

Icon/char for clear search is not loaded, but that no blocker in my eyes

nicodh commented:
> Icon/char for clear search is not loaded, but that no blocker in my eyes

You have an idea why? It's just a Unicode 🞫 ?

Simon-Laux commented:
> You have an idea why? It's just a Unicode 🞫 ?

you need to load a font that supports it

nicodh commented:
> > You have an idea why? It's just a Unicode 🞫 ?
> 
> you need to load a font that supports it

You mean it depends if your system font supports it? Does it makes sense to look for another sign with better support? Maybe https://en.wikipedia.org/wiki/Multiplication_sign ?

Simon-Laux commented:
- either add a font that supports it to the css of the help file
- use just the "X" sign
- use a string/button with the text "clear"
- or use an image/icon.

r10s commented:
this PR has reverted to help to a pretty old state, i'll try and update at https://github.com/deltachat/deltachat-desktop/pull/4725
